### PR TITLE
Add benchmarks for `KedroDataCatalog` and fix tests for `DataCatalog`

### LINKED ISSUE #4125: [Stress Testing] - Data Catalog and Config Loader
## Description

To create stress test for individual components:
- DataCatalog
- ConfigLoader

> As for DataCatalog, the most important thing is to test it within the pipeline, CLI and separately by simulating scenarios when calling some methods as (add_feed_dict). Where the tests themselves should include different sets and combinations of parameters, datasets and patterns. https://github.com/kedro-org/kedro/issues/3957#issuecomment-2299105368

The tickets combined the two components as I believe the setup will be similar, or at least there need to be some mechanism to setup different datasets first so I thought it makes sense to bundle these two components together.

## Context

https://github.com/kedro-org/kedro/issues/3957#issuecomment-2298861806

> ## Component stress test:
> - The main goal for this is to benchmark performance of individual component, this will inform if refactoring work has positive/negative impact. Currently we only check if test pass, so we have no idea if a change may slow down performance. We have done this in the past but usually ad-hoc basis, we should run this regularly (or at least per release).
> 
> The direction of this is simple, we want to make measure the change of time against # number of entries. We would start with Datasets and Catalog, as this fits in the `DataCatalog2.0` work and will be immediately useful.
> 
> 
> - DataCatalog (test # of datasets with catalog.yml & dataset factory)
> - ConfigLoader(# of parameters)
> - Optional: pipelines generated in loops (Dynamic pipeline)
> 
> This can address:
> - https://github.com/kedro-org/kedro/issues/3893
> - DataCatalog 2.0
>

### PR DESCRIPTION
**Pull request by ankatiyar**
## Description

Close #4125 

## Development notes

- Added tests for `KedroDataCatalog`
- Fix tests for `DataCatalog`: The tests were failing for `DataCatalog.add_feed_dict()` and `DataCatalog.add_all()` because asv runs the `setup()` then repeats the tests for a number of times and then runs `teardown()` but trying to add the same datasets multiple times results in `DatasetAlreadyExistsError` https://github.com/kedro-org/kedro/actions/workflows/benchmark-performance.yml

To test locally: `asv run`

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

**Review by merelcht — APPROVED** [review]

File: benchmarks/benchmark_kedrodatacatalog.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    def time_shallow_copy(self):
+        """Benchmark the time to shallow copy the catalog"""
+        # Will be removed
+        self.catalog.shallow_copy()
+
+    def time_resolve_factory(self):
+        """Benchmark the time to resolve factory"""
+        for i in range(1,1001):
+            self.catalog._get_dataset(f"dataset_factory_{i}")
```
Comment (ElenaKhaustova) [inline benchmarks/benchmark_kedrodatacatalog.py:95]: This method will be removed, did you keep it intentionally to compare with the old catalog? If not - you can use `get()` instead

File: benchmarks/benchmark_kedrodatacatalog.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    def time_getitem(self):
+        """Benchmark the time to get a dataset"""
+        for i in range(1,1001):
+            self.catalog[f"dataset_{i}"]
+
+    def time_setitem(self):
+        """Benchmark the time to set a dataset"""
+        for i in range(1,1001):
+            self.catalog[f"dataset_new_{i}"] = CSVDataset(filepath="data.csv")
```
[the remark this answers is not in the document]
Comment (ankatiyar) [reply benchmarks/benchmark_kedrodatacatalog.py:86]: Added a separate test for this

File: benchmarks/benchmark_kedrodatacatalog.py (hunk 1 above)
Comment (ankatiyar) [reply benchmarks/benchmark_kedrodatacatalog.py]: I changed it to use `get()` with the factory ones - i'm assuming this also tests `config_resolver.resolve_pattern()` too

**Review by ElenaKhaustova — APPROVED** [review]
Thank you 🚀

**Pull request merged**